Introduce the precaching post (#1094)
GoogleChrome/web.dev: src/site/content/en/angular/precaching-with-the-angular-service-worker/index.md

---
layout: post
title: 'Precaching with the Angular service worker'
subhead: |

>   Use the Angular service worker to make your app faster and more reliable on networks with poor connectivity!

Reviewer: ```suggestion
  Use the Angular service worker to make your app faster and more reliable on networks with poor connectivity.
```

> hero: performance-budgets-with-the-angular-cli.jpg

Reviewer: ```suggestion
hero: precaching-with-the-angular-service-worker.jpg
alt: A metal grinder throwing sparks.
```
Maybe just rename the image `hero.jpg`? :)

Reviewer: Path is still wrong.

date: 2019-07-02
description: |
  Learn how to use the Angular service worker to precache the static assets in your app.
authors:
  - mgechev
---

## Dealing with limited connectivity

> When users have limited network access—or none at all—web app functionality can significantly degrade and often fails. Using a [service worker](https://developers.google.com/web/fundamentals/primers/service-workers/) to provide precaching lets you intercept network requests and deliver responses directly from a local cache instead of retrieving them from the network. Once your app’s assets have been cached, this approach can really speed up an app and make it work when the user is offline.

Reviewer: ```suggestion
When users have limited network access—or none at all—web app functionality can significantly degrade and often fails. Using a [service worker](https://developers.google.com/web/fundamentals/primers/service-workers/) to provide precaching lets you intercept network requests and deliver responses directly from a local cache instead of retrieving them from the network. Once your app's assets have been cached, this approach can really speed up an app and make it work when the user is offline.
```

> This post walks through how to set up precaching in an Angular app. It assumes you’re already familiar with precaching and service workers in general. If you need a refresher, check out the “[Service workers and the Cache Storage API](https://web.dev/service-workers-cache-storage/)” post.

Reviewer: ```suggestion
This post walks through how to set up precaching in an Angular app. It assumes you're already familiar with precaching and service workers in general. If you need a refresher, check out the "[Service workers and the Cache Storage API](https://web.dev/service-workers-cache-storage/)" post.
```

> _You can find the code for the current example [on GitHub](https://github.com/mgechev/service-worker-web-dev)._

Reviewer: Do we need this since there's an embedded StackBlitz? (If so, suggest putting it in an aside.)

Author: StackBlitz does not demonstrate the service worker in this case.

## Introducing the Angular service worker

> The Angular team offers a service worker module with precaching functionality that’s well integrated with the framework and the [Angular command line interface (CLI)](https://cli.angular.io/).

Reviewer: ```suggestion
The Angular team offers a service worker module with precaching functionality that's well integrated with the framework and the [Angular command line interface (CLI)](https://cli.angular.io/).
```

> Here’s a basic Angular app:

Reviewer: ```suggestion
Here's a basic Angular app:
```

<div class="glitch-embed-wrap" style="height: 480px; width: 100%;">
  <iframe
    src="https://stackblitz.com/github/mgechev/service-worker-web-dev/tree/external-image?embed=1&file=src/app/app.component.ts&view=preview"
    style="height: 100%; width: 100%; border: 0;">
  </iframe>
</div>

> To add the service worker:

Reviewer: ```suggestion
To add the service worker, run this command in the CLI:
```

> 

Reviewer: ```suggestion
```

> 1. Run this command in the CLI:

Reviewer: ```suggestion
```

```

> $ ng add @angular/pwa

Reviewer: I get "Error: Unexpected identifier" when I type this in the StackBlitz console. Also, I think we need some instructions about how to open the StackBlitz console before this instruction.

```

{% Aside %}

If you have multiple projects in the Angular CLI workspace, you can optionally specify a `--project` property with the project name you want to add the service worker to.

{% endAside %}

> `@angular/service-worker` and `@angular/pwa` should now be installed in the app and should appear in `package.json`. The `ng-add` [schematic](https://angular.io/guide/schematics) also adds a file called `ngsw-config.json` to the project, which you can use to configure the service worker. (The file includes a default configuration that you’ll customize a little later.)

Reviewer: ```suggestion
`@angular/service-worker` and `@angular/pwa` should now be installed in the app and should appear in `package.json`. The `ng-add` [schematic](https://angular.io/guide/schematics) also adds a file called `ngsw-config.json` to the project, which you can use to configure the service worker. (The file includes a default configuration that you'll customize a little later.)
```

Now build the project for production:

```
$ ng build --prod
```

> Inside the `dist/[PROJECT_NAME]` directory you’ll find a file called `ngsw.json`. This file tells the Angular service worker how to cache the assets in the app. The file is generated during the build process based on the configuration (`ngsw-config.json`) and the assets produced at build time.

Reviewer: ```suggestion
Inside the `dist/[PROJECT_NAME]` directory you'll find a file called `ngsw.json`. This file tells the Angular service worker how to cache the assets in the app. The file is generated during the build process based on the configuration (`ngsw-config.json`) and the assets produced at build time.
```

> Inside the `dist/[PROJECT_NAME]` directory you’ll find a file called `ngsw.json`. This file tells the Angular service worker how to cache the assets in the app. The file is generated during the build process based on the configuration (`ngsw-config.json`) and the assets produced at build time.

Reviewer: Since we have an actual StackBlitz now, change [PROJECT_NAME] to the actual name in the SB. (I can't see what it should be because commands aren't working in the SB console for me.)

Author: Sounds good.

> Now start an HTTP server in the directory containing your app’s production assets, open the public URL, and check out its network requests in Chrome DevTools:

Reviewer: ```suggestion
Now start an HTTP server in the directory containing your app's production assets, open the public URL, and check out its network requests in Chrome DevTools:
```

> {% Instruction 'preview' %}

Reviewer: This is the shortcode for full-screen view in Glitch. It needs to be replaced with instructions for how to open a new window in StackBlitz.

Reviewer: Since StackBlitz was abandoned, this instruction needs to be update to match what readers will do with the GH version.

> {% Instruction 'devtools-network' %}

Reviewer: ```suggestion
{% Instruction 'devtools-network', 'ol' %}
```
To make this an ordered list.

Note that the network tab has a bunch of static assets directly downloaded in the background by the `ngsw-worker.js` script:

> ![Sample app](sample-app.png "Sample app")

Reviewer: Suggest retaking or editing the screenshot show it shows the StackBlitz URL.

Author: I dropped StackBlitz editor because it doesn't show the service worker integration.

This is the Angular service worker precaching the static assets specified in the generated `ngsw.json` manifest file.

One important asset is missing though: `nyan.png`. To precache this image you need to add a pattern that includes it to `ngsw-config.json`, which lives in the root of the workspace:

```json/13
{
  "$schema": "./node_modules/@angular/service-worker/config/schema.json",
  "index": "/index.html",
  "assetGroups": [
    {
      "name": "app",
      "installMode": "prefetch",
      "resources": {
        "files": [
          "/favicon.ico",
          "/index.html",
          "/*.css",
          "/*.js",
          "/assets/*.png"
        ]
      }
    },
    ...
}
```

This change adds all PNG images in the `/assets` folder  to the `app` resource asset group. Since the `installMode` for this asset group is set to `prefetch`, the service worker will precache all the specified assets—which now include PNG images.

Specifying other assets to be precached is just as straightforward: update the patterns in the `app` resource asset group.

## Conclusion

> Using a service worker for precaching can improve the performance of your apps by saving assets to a local cache and make them more reliable on poor networks. To use precaching with Angular and the Angular CLI:

Reviewer: ```suggestion
Using a service worker for precaching can improve the performance of your apps by saving assets to a local cache, which makes them more reliable on poor networks. To use precaching with Angular and the Angular CLI:
```

> 1. Add the `@angular/pwa` package to your project

Reviewer: ```suggestion
1. Add the `@angular/pwa` package to your project.
```

> 2. Control what the service worker caches by editing `ngsw-config.js`

Reviewer: ```suggestion
2. Control what the service worker caches by editing `ngsw-config.js`.
```
